File: frontend/src/portfolio-import/instrumentRegistry.ts
/**
 * Portfolio Import v1.2 — bounded, explicit instrument registry.
 *
 * Replaces v1.1's flat name -> ticker dictionary. Still deterministic,
 * still a maintained lookup, not an attempt at universal coverage or a
 * general security master — the point is identity *safety* for this one
 * import flow: every entry represents one specific, verified instrument
 * identity, not a guess.
 *
 * Two safety rules this registry exists to enforce, both learned from a
 * real portfolio import (v1.2's hardening pass):
 *
 * 1. A company with more than one actively-traded public share class
 *    (Alphabet, Berkshire Hathaway) has NO bare/unqualified entry here.
 *    Only the fully-qualified name ("Alphabet Class C") resolves —
 *    defaulting an unqualified name to one class would be exactly the
 *    kind of guessed suffix this sprint forbids. An investor who pastes
 *    just "Alphabet" is left to confirm which class manually.
 * 2. An instrument whose identity is known but isn't a plain listed
 *    equity (a fund, an ETP, a private company) still gets an entry —
 *    with `ticker: null` — so the review screen can say "this is a
 *    recognized fund" instead of either fabricating a stock ticker for
 *    it or treating it as a blank unknown.
 *
 * Nordic-only local listings the real portfolio contained (Volvo B,
 * Alfa Laval, Trelleborg B, Sandvik, Assa Abloy B, Atlas Copco B,
 * SEB A, Schneider Electric) are deliberately left OUT of this
 * registry rather than mapped to a guessed Nasdaq Stockholm / Euronext
 * suffix convention Atlas has never verified — they fall through to
 * manual confirmation, per this sprint's own instruction to "verify
 * each mapping... or leave it unresolved."
 */

export type InstrumentType = "equity" | "fund" | "etp" | "private" | "other";

export interface InstrumentRegistryEntry {
  /** Names an investor would plausibly type or paste, lowercase-normalized
   *  for lookup. Not fuzzy aliases — exact known spellings only. */
  displayNames: string[];
  /** `null` when this instrument is recognized but isn't a plain listed
   *  equity the current Alpha backend can honestly persist as ticker +
   *  weight — see `resolution.ts`'s "unsupported" resolution kind. */
  ticker: string | null;
  instrumentType: InstrumentType;
  market?: string;
  shareClass?: string;
}

export const INSTRUMENT_REGISTRY: readonly InstrumentRegistryEntry[] = [
  // ---- unambiguous US/NASDAQ/NYSE equities, single public class ----
  { displayNames: ["microsoft"], ticker: "MSFT", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["apple"], ticker: "AAPL", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["amazon", "amazon.com"], ticker: "AMZN", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["nvidia"], ticker: "NVDA", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["tesla"], ticker: "TSLA", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["broadcom"], ticker: "AVGO", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["visa"], ticker: "V", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["mastercard"], ticker: "MA", instrumentType: "equity", market: "NYSE" },
  {
    displayNames: ["johnson & johnson", "johnson and johnson"],
    ticker: "JNJ",
    instrumentType: "equity",
    market: "NYSE",
  },
  {
    displayNames: ["procter & gamble", "procter and gamble"],
    ticker: "PG",
    instrumentType: "equity",
    market: "NYSE",
  },
  { displayNames: ["exxon mobil", "exxonmobil"], ticker: "XOM", instrumentType: "equity", market: "NYSE" },
  {
    displayNames: ["jpmorgan chase", "jpmorgan", "jp morgan"],
    ticker: "JPM",
    instrumentType: "equity",
    market: "NYSE",
  },
  { displayNames: ["walmart"], ticker: "WMT", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["eli lilly"], ticker: "LLY", instrumentType: "equity", market: "NYSE" },
  {
    displayNames: ["unitedhealth group", "unitedhealth"],
    ticker: "UNH",
    instrumentType: "equity",
    market: "NYSE",
  },
  { displayNames: ["home depot"], ticker: "HD", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["salesforce"], ticker: "CRM", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["adobe"], ticker: "ADBE", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["netflix"], ticker: "NFLX", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["coca-cola", "coca cola"], ticker: "KO", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["pepsico"], ticker: "PEP", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["costco"], ticker: "COST", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["oracle"], ticker: "ORCL", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["intel"], ticker: "INTC", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["cisco"], ticker: "CSCO", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["qualcomm"], ticker: "QCOM", instrumentType: "equity", market: "NASDAQ" },
  { displayNames: ["texas instruments"], ticker: "TXN", instrumentType: "equity", market: "NASDAQ" },
  {
    displayNames: ["thermo fisher scientific"],
    ticker: "TMO",
    instrumentType: "equity",
    market: "NYSE",
  },
  { displayNames: ["abbvie"], ticker: "ABBV", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["merck"], ticker: "MRK", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["pfizer"], ticker: "PFE", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["verizon"], ticker: "VZ", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["at&t"], ticker: "T", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["disney", "walt disney"], ticker: "DIS", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["nike"], ticker: "NKE", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["mcdonald's", "mcdonalds"], ticker: "MCD", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["starbucks"], ticker: "SBUX", instrumentType: "equity", market: "NASDAQ" },
  {
    displayNames: ["ibm", "international business machines"],
    ticker: "IBM",
    instrumentType: "equity",
    market: "NYSE",
  },
  { displayNames: ["paypal"], ticker: "PYPL", instrumentType: "equity", market: "NASDAQ" },
  {
    displayNames: ["booking holdings", "booking.com"],
    ticker: "BKNG",
    instrumentType: "equity",
    market: "NASDAQ",
  },
  { displayNames: ["uber"], ticker: "UBER", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["spotify"], ticker: "SPOT", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["shopify"], ticker: "SHOP", instrumentType: "equity", market: "NYSE" },
  {
    displayNames: ["meta platforms", "meta platforms a", "meta"],
    ticker: "META",
    instrumentType: "equity",
    market: "NASDAQ",
    shareClass: "A",
  },
  { displayNames: ["vistra"], ticker: "VST", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["vertiv"], ticker: "VRT", instrumentType: "equity", market: "NYSE" },
  { displayNames: ["applied materials"], ticker: "AMAT", instrumentType: "equity", market: "NASDAQ" },

  // ---- multi-class companies: only the fully-qualified name resolves;
  // the bare/unqualified name is deliberately absent from this registry
  // so it falls through to manual confirmation instead of defaulting
  // to a guessed class. ----
  {
    displayNames: ["alphabet class a"],
    ticker: "GOOGL",
    instrumentType: "equity",
    market: "NASDAQ",
    shareClass: "A",
  },
  {
    displayNames: ["alphabet class c"],
    ticker: "GOOG",
    instrumentType: "equity",
    market: "NASDAQ",
    shareClass: "C",
  },
  {
    displayNames: ["berkshire hathaway class a"],
    ticker: "BRK.A",
    instrumentType: "equity",
    market: "NYSE",
    shareClass: "A",
  },
  {
    displayNames: ["berkshire hathaway class b"],
    ticker: "BRK.B",
    instrumentType: "equity",
    market: "NYSE",
    shareClass: "B",
  },

  // ---- non-US primary listings represented by an unambiguous ADR ----
  { displayNames: ["taiwan semiconductor"], ticker: "TSM", instrumentType: "equity", market: "NYSE (ADR)" },
  { displayNames: ["astrazeneca"], ticker: "AZN", instrumentType: "equity", market: "NASDAQ (ADR)" },
  {
    displayNames: ["novo nordisk", "novo nordisk b"],
    ticker: "NVO",
    instrumentType: "equity",
    market: "NYSE (ADR)",
    shareClass: "B",
  },

  // ---- non-equity instruments: identity is known, but the current
  // Alpha backend can only persist ticker + weight and none of these
  // are a plain listed equity. `ticker` is deliberately `null` —
  // resolution.ts surfaces these as "unsupported", never a fabricated
  // stock ticker. ----
  {
    displayNames: ["coinshares xbt provider bitcoin tracker one", "coinshares xbt provider"],
    ticker: null,
    instrumentType: "etp",
  },
  {
    displayNames: ["länsförsäkringar global index", "lansforsakringar global index"],
    ticker: null,
    instrumentType: "fund",
  },
  { displayNames: ["avanza emerging markets"], ticker: null, instrumentType: "fund" },
  { displayNames: ["spacex"], ticker: null, instrumentType: "private" },
];

const LOOKUP: ReadonlyMap<string, InstrumentRegistryEntry> = new Map(
  INSTRUMENT_REGISTRY.flatMap((entry) => entry.displayNames.map((name) => [name, entry] as const)),
);

export function lookupInstrument(name: string): InstrumentRegistryEntry | null {
  return LOOKUP.get(name.trim().toLowerCase()) ?? null;
}

File: atlas/alpha/portfolio_import/instrument_registry.py
"""Instrument Identity registry, ported from the frontend's own
`frontend/src/portfolio-import/instrumentRegistry.ts` -- the same
bounded, explicit, deterministic entries, not an attempt at universal
coverage. See that file's own module docstring for the full rationale
(no bare entry for a multi-share-class company, `ticker=None` for a
known-but-unsupported instrument type, only high-confidence Nordic
tickers). Kept as a second copy rather than a shared source for this
sprint -- the frontend's own copy is retired once the new unified
onboarding screen ships (see the Zero-Effort Onboarding Architecture's
Implementation Plan, Phase 5); consolidating the registry into one
place afterward is a natural, small follow-up, not blocking this one.
"""
from __future__ import annotations

import re
from dataclasses import dataclass

InstrumentType = str  # "equity" | "fund" | "etp" | "private" | "other"


@dataclass(frozen=True)
class InstrumentRegistryEntry:
    display_names: tuple[str, ...]
    ticker: str | None
    instrument_type: InstrumentType


INSTRUMENT_REGISTRY: tuple[InstrumentRegistryEntry, ...] = (
    InstrumentRegistryEntry(("microsoft", "microsoft corp", "microsoft corporation"), "MSFT", "equity"),
    InstrumentRegistryEntry(("apple",), "AAPL", "equity"),
    InstrumentRegistryEntry(("amazon", "amazon.com", "amazon.com inc"), "AMZN", "equity"),
    InstrumentRegistryEntry(("nvidia",), "NVDA", "equity"),
    InstrumentRegistryEntry(("tesla",), "TSLA", "equity"),
    InstrumentRegistryEntry(("broadcom",), "AVGO", "equity"),
    InstrumentRegistryEntry(("visa",), "V", "equity"),
    InstrumentRegistryEntry(("mastercard",), "MA", "equity"),
    InstrumentRegistryEntry(("johnson & johnson", "johnson and johnson"), "JNJ", "equity"),
    InstrumentRegistryEntry(("procter & gamble", "procter and gamble"), "PG", "equity"),
    InstrumentRegistryEntry(("exxon mobil", "exxonmobil"), "XOM", "equity"),
    InstrumentRegistryEntry(("jpmorgan chase", "jpmorgan", "jp morgan"), "JPM", "equity"),
    InstrumentRegistryEntry(("walmart",), "WMT", "equity"),
    InstrumentRegistryEntry(("eli lilly",), "LLY", "equity"),
    InstrumentRegistryEntry(("unitedhealth group", "unitedhealth"), "UNH", "equity"),
    InstrumentRegistryEntry(("home depot",), "HD", "equity"),
    InstrumentRegistryEntry(("salesforce",), "CRM", "equity"),
    InstrumentRegistryEntry(("adobe",), "ADBE", "equity"),
    InstrumentRegistryEntry(("netflix",), "NFLX", "equity"),
    InstrumentRegistryEntry(("coca-cola", "coca cola"), "KO", "equity"),
    InstrumentRegistryEntry(("pepsico",), "PEP", "equity"),
    InstrumentRegistryEntry(("costco",), "COST", "equity"),
    InstrumentRegistryEntry(("oracle",), "ORCL", "equity"),
    InstrumentRegistryEntry(("intel",), "INTC", "equity"),
    InstrumentRegistryEntry(("cisco",), "CSCO", "equity"),
    InstrumentRegistryEntry(("qualcomm",), "QCOM", "equity"),
    InstrumentRegistryEntry(("texas instruments",), "TXN", "equity"),
    InstrumentRegistryEntry(("thermo fisher scientific",), "TMO", "equity"),
    InstrumentRegistryEntry(("abbvie",), "ABBV", "equity"),
    InstrumentRegistryEntry(("merck",), "MRK", "equity"),
    InstrumentRegistryEntry(("pfizer",), "PFE", "equity"),
    InstrumentRegistryEntry(("verizon",), "VZ", "equity"),
    InstrumentRegistryEntry(("at&t",), "T", "equity"),
    InstrumentRegistryEntry(("disney", "walt disney"), "DIS", "equity"),
    InstrumentRegistryEntry(("nike",), "NKE", "equity"),
    InstrumentRegistryEntry(("mcdonald's", "mcdonalds"), "MCD", "equity"),
    InstrumentRegistryEntry(("starbucks",), "SBUX", "equity"),
    InstrumentRegistryEntry(("ibm", "international business machines"), "IBM", "equity"),
    InstrumentRegistryEntry(("paypal",), "PYPL", "equity"),
    InstrumentRegistryEntry(("booking holdings", "booking.com"), "BKNG", "equity"),
    InstrumentRegistryEntry(("uber",), "UBER", "equity"),
    InstrumentRegistryEntry(("spotify",), "SPOT", "equity"),
    InstrumentRegistryEntry(("shopify",), "SHOP", "equity"),
    InstrumentRegistryEntry(
        ("meta platforms", "meta platforms a", "meta platforms inc", "meta"), "META", "equity"
    ),
    InstrumentRegistryEntry(("vistra",), "VST", "equity"),
    InstrumentRegistryEntry(
        ("vertiv", "vertiv holdings", "vertiv holdings a", "vertiv holdings co"), "VRT", "equity"
    ),
    InstrumentRegistryEntry(("applied materials",), "AMAT", "equity"),
    # multi-class companies: only the fully-qualified name resolves.
    InstrumentRegistryEntry(("alphabet class a", "alphabet inc class a"), "GOOGL", "equity"),
    InstrumentRegistryEntry(("alphabet class c", "alphabet inc class c"), "GOOG", "equity"),
    InstrumentRegistryEntry(
        ("berkshire hathaway class a", "berkshire hathaway inc class a"), "BRK.A", "equity"
    ),
    InstrumentRegistryEntry(
        ("berkshire hathaway class b", "berkshire hathaway inc class b"), "BRK.B", "equity"
    ),
    # non-US primary listings represented by an unambiguous ADR.
    InstrumentRegistryEntry(
        (
            "taiwan semiconductor",
            "taiwan semiconductor mfg co",
            "taiwan semiconductor manufacturing",
            "tsmc",
        ),
        "TSM",
        "equity",
    ),
    InstrumentRegistryEntry(("astrazeneca",), "AZN", "equity"),
    InstrumentRegistryEntry(("novo nordisk", "novo nordisk b"), "NVO", "equity"),
    # Nasdaq Stockholm large caps -- only high-confidence local tickers.
    InstrumentRegistryEntry(("investor b", "investor ab b", "investor ab class b"), "INVE-B", "equity"),
    InstrumentRegistryEntry(("atlas copco b", "atlas copco ab b"), "ATCO-B", "equity"),
    InstrumentRegistryEntry(("volvo b", "volvo ab b"), "VOLV-B", "equity"),
    InstrumentRegistryEntry(("assa abloy b", "assa abloy ab b"), "ASSA-B", "equity"),
    InstrumentRegistryEntry(("seb a", "seb ab a"), "SEB-A", "equity"),
    InstrumentRegistryEntry(("alfa laval",), "ALFA", "equity"),
    InstrumentRegistryEntry(("sandvik",), "SAND", "equity"),
    InstrumentRegistryEntry(("trelleborg b", "trelleborg ab b"), "TREL-B", "equity"),
    InstrumentRegistryEntry(("abb", "abb ltd"), "ABB", "equity"),
    InstrumentRegistryEntry(("munters", "munters group", "munters group ab"), "MTRS", "equity"),
    InstrumentRegistryEntry(("castellum",), "CAST", "equity"),
    InstrumentRegistryEntry(("nordnet",), "SAVE", "equity"),
    # non-equity instruments: identity known, ticker deliberately None.
    InstrumentRegistryEntry(
        ("coinshares xbt provider bitcoin tracker one", "coinshares xbt provider"), None, "etp"
    ),
    InstrumentRegistryEntry(
        ("länsförsäkringar global index", "lansforsakringar global index"), None, "fund"
    ),
    InstrumentRegistryEntry(("avanza emerging markets",), None, "fund"),
    InstrumentRegistryEntry(("spacex",), None, "private"),
)

_DASH_VARIANT_PATTERN = re.compile("[‐‑‒–—−]")
_TRAILING_PERIOD_PATTERN = re.compile(r"\.(?=\s|$)")
_WHITESPACE_PATTERN = re.compile(r"\s+")


def normalize_for_lookup(raw: str) -> str:
    """Punctuation/whitespace normalization only, never a matching
    strategy on its own -- ported verbatim from `instrumentRegistry.ts`'s
    identical function."""
    text = _DASH_VARIANT_PATTERN.sub("-", raw)
    text = text.replace(",", " ")
    text = _TRAILING_PERIOD_PATTERN.sub("", text)
    text = _WHITESPACE_PATTERN.sub(" ", text)
    return text.strip().lower()


_LOOKUP: dict[str, InstrumentRegistryEntry] = {
    normalize_for_lookup(name): entry
    for entry in INSTRUMENT_REGISTRY
    for name in entry.display_names
}


def lookup_instrument(name: str) -> InstrumentRegistryEntry | None:
    return _LOOKUP.get(normalize_for_lookup(name))
